include/boost/math/special_functions/chebyshev.hpp

```diff
@@ -0,0 +1,96 @@
+//  (C) Copyright Nick Thompson 2017.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_CHEBYSHEV_HPP
+#define BOOST_MATH_SPECIAL_CHEBYSHEV_HPP
+#include <cmath>
+
+namespace boost { namespace math {
+
+template<class Real>
+inline Real chebyshev_next(Real const & x, Real const & Tn, Real const & Tn_1)
+{
+    return 2*x*Tn - Tn_1;
+}
+
+namespace detail {
+
+template<class Real, bool second=false>
+inline Real chebyshev_imp(unsigned n, Real const & x)
+{
+    using std::cosh;
+    using std::acosh;
+    using std::pow;
+    Real T0 = 1;
+    Real T1;
+    if (second)
```

Comment:
To be honest, I think I'd rather have 2 separate functions than switch based on a template parameter - there's not much overlap between the two cases other than the recursion at the end?